File: src/main/java/com/bk/Day_1_ReportRepair.java
package com.bk;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;

public class Day_1_ReportRepair {

    public void part1() throws Exception {
        Path path = Paths.get(getClass().getClassLoader().getResource("input.txt").toURI());
        List<String> lines = Files.readAllLines(path);
        int[] report = lines.stream().mapToInt(Integer::valueOf).toArray();

        Arrays.sort(report);    // worst-case sort => O(n log n)

        int a = 0;
        int b = report.length - 1;
        int sum, val1, val2;

        while (a < b) {
            val1 = report[a];
            val2 = report[b];
            sum = val1 + val2;
            if (sum == 2020) {
                System.out.printf("input = %s, val1 = %d, val2 = %d, product = %d", Arrays.toString(report), val1, val2, val1 * val2);
                return;
            } else if (sum < 2020) {
                // sum too low - move 1st pointer to next highest
                a++;
            } else {
                // only possibility here is sum > 2020, so sum is too high - move 2nd pointer to next lowest
                b--;
            }
        }
    }

    public static void main(String[] args) throws Exception {
        Day_1_ReportRepair inst = new Day_1_ReportRepair();
        inst.part1();
    }
}

File: src/python/day_1_report_repair.py
#!/usr/bin/env python3

def part1(input_list):
    expenses = sorted(input_list)

    a = 0
    b = len(expenses) - 1

    while a < b:
        val1 = expenses[a]
        val2 = expenses[b]
        sum = val1 + val2
        if sum == 2020:
            print("input = {}, val1 = {}, val2 = {}, product = {}".format(input_list, val1, val2, val1 * val2))
            return
        elif sum < 2020:
            # sum too low - move 1st pointer to next highest
            a += 1
        else:
            # only possibility here is sum > 2020, so sum is too high - move 2nd pointer to next lowest
            b -= 1


if __name__ == '__main__':
    file = open('../main/resources/input.txt', mode="r")
    ints = []
    for s in file:
        ints.append(int(s))

    part1(ints)

    # part1([1721, 979, 366, 299, 675, 1456])
    # part1([10, 20, 1009, 4000, 3000, 1011])
